### ISSUE TITLE
Batching loss function over modules

### ISSUE THREAD

**Issue by JamesRoggeveen** [patrick-kidger/equinox#1079]
I find myself trying to write a program where it is convenient to have a container class for data, which is then passed into a loss function as part of an optimization loop. In some cases, we may have multiple data sets that we would like to optimize over at the same time (i.e. the total loss is the sum of the loss applied to each data class). However, these data sets may intrinsically have different sizes, meaning we cannot simply `vmap` over the container class. I tried to write a batch class that holds a collection of these data classes and use `eqx.filter_vmap` on this container class, but continue to run into the issue that these jax transformations want to apply to each leaf of the pytree, rather than only going one level down (I tried `jax.lax.map` as well). What is the preferred Equinox approach to solving this problem? I have included a minimum working example below, including fields that would be used in a properly implemented loss function.
```
import jax.numpy as jnp
import jax
import equinox as eqx

class DataStorage(eqx.Module):
    data: jax.Array
    time: jax.Array
    condition: jax.Array

class BatchedData(eqx.Module):
    data: tuple[DataStorage, ...]

def sample_loss(data):
    return jnp.mean(data.data**2)

data_1 = DataStorage(data=jnp.array([1,2,3]), time=jnp.array([1,2,3]), condition=jnp.array([1,2]))
data_2 = DataStorage(data=jnp.array([4,5,6,7,8,9]), time=jnp.array([1,2,3,4,5,6]), condition=jnp.array([1,3]))
data_3 = DataStorage(data=jnp.array([10,11,12,13,14,15]), time=jnp.array([1,2,3,4,5,6]), condition=jnp.array([2,3]))

batched_data = BatchedData(data=(data_1, data_2, data_3))

eqx.filter_vmap(sample_loss)(batched_data)
```

**Comment by lockwo** [comment]
I'm not sure there's really an equinox solution to this, this is a sort of a JAX thing. Jax likes vmap array shapes to be the same (more specifically, it just needs arrays to vmap over and since it doesn't have ragged arrays, it naturally requires same shapes). In this case equinox is just providing a nice wrapper for pytrees, but cut equinox out of the picture and the problem is the same: you want to vmap over arrays that are not the same size (regardless of the complexity or structure of the pytree, that is the crux of it). Unfortunately, equinox doesn't have any miracle cures for this (although I am always on the hunt for them). 

When I've encountered this before I've done two things. First is to pad. If the array sizes are massively different, you can pad the arrays to the largest size and also provide a mask, then just multiply by the mask inside the computations, then do the computation (taking care the padding doesn't effect anything). Second is to just iterate. Sometimes this can avoid redundant computation and can be a fine solution.

**Comment by michael-0brien** [comment]
I’m not sure I fully understand what you’re trying to do with this vmap, but depending on what that is the discussion in #618 could be useful.
